candle-nn/src/batch_norm.rs

```diff
@@ -142,26 +184,51 @@ impl BatchNorm {
         let x = x.flatten_from(1)?.contiguous()?;
         let x = if self.remove_mean {
             let mean_x = x.mean_keepdim(1)?;
+            {
+                // Update running mean
+                let new_mean = ((self.running_mean.as_tensor() * (1.0 - self.momentum))?
+                    + (mean_x.reshape(((),))? * self.momentum)?)?;
+
+                self.running_mean.set(&new_mean)?;
+            }
             x.broadcast_sub(&mean_x)?
         } else {
             x
         };
         let norm_x = x.sqr()?.mean_keepdim(1)?;
+        {
+            // Update running variance
+            let running_var_weight = 1.0 - self.momentum;
+            let norm_x_weight = self.momentum * num_features as f64 / (num_features as f64 - 1.0);
```

Comment:
I'm a bit confused by this, shouldn't this be the batch-size rather than the number of features to get the unbiased estimator?
But that would also mean that the two `mean_keepdim` above should be done on the first dimension (batch dim) rather than on the second (channels)?
What testing did you run on this?

Reply:
This should be resolved now, and I have added to the batchnorm test to verify the new behavior against pytorch.